NIFI-14375 - Kafka refactor - partitioning, uncommitted time, inputstream, tests, etc

pvillard31 opened the pull request:
# Summary

[NIFI-14375](https://issues.apache.org/jira/browse/NIFI-14375) - Kafka refactor - partitioning, uncommitted time, inputstream, tests, etc

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/kafka/processors/ConsumeKafka.java:
```diff
@@ -365,23 +363,53 @@ public void onStopped() {
+                    getLogger().error("Failed to consume Kafka Records", e);
+                    consumerService.rollback();
+
+                    try {
+                        consumerService.close();
+                    } catch (final IOException ex) {
+                        getLogger().warn("Failed to close Kafka Consumer Service", ex);
+                    }
```
You can use try with resources to close `consumerService` although you will end up with an empty `try` block.  Ditto for lines 406-410
```suggestion
                           try (consumerService){
            } catch (IOException ex) {
                getLogger().warn("Failed to close Kafka Consumer Service", ex);
            }
```

pvillard31 replied:
I feel like explicitly calling the `close()` here is making things easier to read compared to an empty try block. I can definitely make the change if there is a consensus that the empty block is better.

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-3-integration/src/test/java/org/apache/nifi/kafka/processors/ConsumeKafkaRecordIT.java:
```diff
@@ -78,6 +85,45 @@ void testProcessingStrategyRecordNoRecords() {
         runner.assertAllFlowFilesTransferred(ConsumeKafka.SUCCESS, 0);
     }
 
+    @Test
+    void testInvalidRecordInMiddle() throws ExecutionException, InterruptedException {
+        testSingleInvalidRecord("testInvalidRecordInMiddle", VALID_RECORD_1_TEXT, INVALID_RECORD_TEXT, VALID_RECORD_2_TEXT);
+    }
+
+    @Test
+    void testInvalidRecordAtEnd() throws ExecutionException, InterruptedException {
+        testSingleInvalidRecord("testInvalidRecordAtEnd", VALID_RECORD_1_TEXT, VALID_RECORD_2_TEXT, INVALID_RECORD_TEXT);
+    }
+
+    @Test
+    void testInvalidRecordAtStart() throws ExecutionException, InterruptedException {
+        testSingleInvalidRecord("testInvalidRecordAtStart", INVALID_RECORD_TEXT, VALID_RECORD_1_TEXT, VALID_RECORD_2_TEXT);
+    }
```
Instead of adding three individual tests create one JUnit5 `ParamaterizedTest`

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-3-integration/src/test/java/org/apache/nifi/kafka/processors/PublishKafkaRoundRobinIT.java:
```diff
@@ -0,0 +1,79 @@
+        int totalRecords = 0;
+        final Map<Integer, Integer> recordsPerPartition = new HashMap<>();
+        try (final KafkaConsumer<String, String> consumer = new KafkaConsumer<>(getKafkaConsumerProperties())) {
+            consumer.subscribe(Collections.singletonList(getClass().getName()));
```
```suggestion
            consumer.subscribe(List.of(getClass().getName()));
```

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-3-service/src/main/java/org/apache/nifi/kafka/service/Kafka3ConnectionService.java:
```diff
@@ -249,15 +252,33 @@ public class Kafka3ConnectionService extends AbstractControllerService implement
+        final String[] bootstrapServersArray = bootstrapServers.split(",");
+        String firstBootstrapServer = bootstrapServersArray[0].trim();
+        if (firstBootstrapServer.contains("://")) {
+            firstBootstrapServer = firstBootstrapServer.substring(firstBootstrapServer.indexOf("://") + 3);
```
I am sorry for my ignorance here regarding Kafka configuration but will all the configured servers always have `://`. The way I read th Kafka documentation `host:port` where host may not include the leading `://`.

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-3-service/src/main/java/org/apache/nifi/kafka/service/consumer/Kafka3ConsumerService.java:
```diff
@@ -28,39 +29,58 @@
+    @Override
+    public void onPartitionsRevoked(final Collection<TopicPartition> partitions) {
+        componentLog.info("Kafka revoked the following Partitions from this consumer: {}", partitions);
+        rollback(new HashSet<>(partitions));
```
Couldn't `Set.of` be sufficient?
```suggestion
        rollback(Set.of(partitions));
```

pvillard31 replied:
I could use `Set.copyOf()` but not sure to see the benefit

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-3-service/src/test/java/org/apache/nifi/kafka/service/Kafka3ConnectionServiceBaseIT.java:
```diff
@@ -267,7 +265,7 @@ void testProduceConsumeRecord() {
         assertNotNull(summary);
 
-        final PollingContext pollingContext = new PollingContext(GROUP_ID, Collections.singleton(TOPIC), AutoOffsetReset.EARLIEST, Duration.ofSeconds(1));
+        final PollingContext pollingContext = new PollingContext(GROUP_ID, Collections.singleton(TOPIC), AutoOffsetReset.EARLIEST);
```
```suggestion
        final PollingContext pollingContext = new PollingContext(GROUP_ID, Set.of(TOPIC), AutoOffsetReset.EARLIEST);
```

dan-s1 commented on nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/test/java/org/apache/nifi/kafka/processors/producer/convert/KafkaRecordConverterTest.java:
```diff
@@ -0,0 +1,155 @@
+        System.arraycopy(stringBytes, 0, LARGE_SAMPLE_INPUT, 0, copyLength);
+    }
+
+    private void verifyRecordCount(Iterator<KafkaRecord> records, int expectedCount) {
+        List<KafkaRecord> recordList = new ArrayList<>();
+        records.forEachRemaining(recordList::add);
+        assertEquals(expectedCount, recordList.size(), "Expected exactly " + expectedCount + " records");
+    }
```
This method and the following 4 tests should all be part of a `ParamaterizedTest`

pvillard31 commented:
Thanks for the review @dan-s1 - I've addressed some of your comments. I didn't apply some recommended changes as I'm not sure this would improve readability / performances. For the parameterized tests, I'm open to the idea but could not find a nice way to make it work with the runner, if you have a working approach, I'm ok with it :)

dan-s1 commented:
Take a look at #9785 where there is an example of one.

pvillard31 commented:
Thanks @dan-s1 - managed to get something working - not sure it makes things easier to understand for a few lines of code gain :)

pvillard31 commented:
Note to reviewers: I just added a commit (https://github.com/apache/nifi/pull/9807/commits/c40cde6df1e5df8f33eebfa18452b9f5b764d658) that properly handles [KIP-794](https://github.com/apache/nifi/pull/9807/commits/c40cde6df1e5df8f33eebfa18452b9f5b764d658) which changes the approach for the default partitioning. This change is required in anticipation of bumping the Kafka client to 4.0.0 where the class is completely removed.

exceptionfactory reviewed (approved):
Thanks again @pvillard31! +1 merging
